lib/vm/src/instance.rs

```diff
@@ -761,110 +661,272 @@ impl Instance {
+/// shared an `Instance` between an `InstanceHandle` and the module
+/// exports, so that one can drop a `InstanceHandle` but still being
+/// able to use the exports properly.
+///
+/// This structure has a C representation because `Instance` is
+/// dynamically-sized, and the `instance` field must be last.
+#[derive(Debug)]
+#[repr(C)]
+pub struct InstanceAllocator {
+    /// Number of `Self` in the nature. It increases when `Self` is
+    /// cloned, and it decreases when `Self` is dropped.
+    strong: Arc<atomic::AtomicUsize>,
+
+    /// The layout of `Instance` (which can vary).
+    instance_layout: Layout,
 
-    /// Whether `Self` owns `self.instance`. It's not always the case;
-    /// e.g. when `Self` is built with `Self::from_vmctx`.
+    /// The `Instance` itself. It must be the last field of
+    /// `InstanceAllocator` since `Instance` is dyamically-sized.
     ///
-    /// This information is necessary to know whether `Self` should
-    /// deallocate `self.instance`.
-    owned_instance: bool,
+    /// `Instance` must not be dropped manually by Rust, because it's
+    /// allocated manually with `alloc` and a specific layout (Rust
+    /// would be able to drop `Instance` itself but it will imply a
+    /// memory leak because of `alloc`).
+    ///
+    /// No one in the code has a copy of the `Instance`'s
+    /// pointer. `Self` is the only one.
+    instance: NonNull<Instance>,
 }
 
-/// # Safety
-/// This is safe because there is no thread-specific logic in `InstanceHandle`.
-/// TODO: this needs extra review
-unsafe impl Send for InstanceHandle {}
+impl InstanceAllocator {
+    /// Create a new `InstanceAllocator`. It allocates nothing. It
+    /// fills nothing. The `Instance` must be already valid and
+    /// filled. `self_ptr` and `self_layout` must be the pointer and
+    /// the layout returned by `Self::allocate_self` used to build
+    /// `Self`.
+    fn new(instance: NonNull<Instance>, instance_layout: Layout) -> Self {
+        Self {
+            strong: Arc::new(atomic::AtomicUsize::new(1)),
+            instance_layout,
+            instance,
+        }
+    }
 
-impl InstanceHandle {
-    /// Allocates an instance for use with `InstanceHandle::new`.
-    ///
-    /// Returns the instance pointer and the [`VMOffsets`] that describe the
-    /// memory buffer pointed to by the instance pointer.
-    pub fn allocate_instance(module: &ModuleInfo) -> (NonNull<u8>, VMOffsets) {
-        let offsets = VMOffsets::new(mem::size_of::<*const u8>() as u8, module);
+    /// Calculate the appropriate layout for `Instance`.
+    fn instance_layout(offsets: &VMOffsets) -> Layout {
+        let size = mem::size_of::<Instance>()
+            .checked_add(
+                usize::try_from(offsets.size_of_vmctx())
+                    .expect("Failed to convert the size of `vmctx` to a `usize`"),
+            )
+            .expect("Failed to compute the size of `Instance`");
+        let align = mem::align_of::<Instance>();
```

Comment:
Actually, it's from the old code base, but I believe it is wrong. It works only because the size of `Instance` + the size of `VMContext` still fits in the alignment of `Instance` only.

I can use a drafted code I developed when working on this patch to compute a more change-proof calculation for the alignment.

Reply:
I think we should do something like this:

```rust
fn instance_Layout(offsets: &VMOffsets) -> Layout {
        let (layout, _offset) = Layout::new::<Instance>()
            .extend(
                Layout::array::<u8>(
                    usize::try_from(offsets.size_of_vmctx())
                        .expect("Failed to convert the size of `vmctx` to a `usize`"),
                )
                .expect("Failed to create a layout for `VMContext`"),
            )
            .expect("Failed to extend to `Instance` layout to include `VMContext`");

        dbg!(layout.pad_to_align())
}
```

On my machine, it gives:

```
Layout {
    size_: 400,
    align_: 16,
}
```

which is correct.

Thoughts?

Reply:
Hmm, actually the size of `VMContext` is given by `VMOffsets.size_of_vmctx` and I'm even not sure it computes the size correctly. Digging.

Reply:
I can't decide whether we should use `Layout::array::<u8>` or `Layout::array::<usize>`. For me, `VMContext` contains pointers, so it should be an array of `usize`, but I'm not sure. Using `u8` gives the same results as we have on `master`, but it may be wrong!

Reply:
Let's keep `u8` for the moment. As you said earlier “Somewhat unrelated”. https://github.com/wasmerio/wasmer/pull/1837/commits/0ec3b70b2267079e4f3ac84fd24b8cf465b0e6d8 already improves how the real `Instance` layout is calculated, which is good, let's keep things as they are.

Reply:
That seems reasonable to me, the one thing I'd add is that I think we need to make sure we have proper padding (to force proper alignment) between the known struct and the dynamic extension of it.

Just because the struct itself is aligned to 16 doesn't mean that the dynamic part will be...

Reply:
I guess `Layout` already handles that case. The dynamic part is handled by `Layout::array::<u8>` which is the extension of `Layout::new::<Instance>`.

In anyways, the actual code is _not_ worst than what we had before ;-).